File: src/snabl/task.cpp
#include "snabl/env.hpp"
#include "snabl/task.hpp"

namespace snabl {
  Task::Task(Env &env, PC start_pc, const ScopePtr &parent_scope):
    scope(parent_scope), prev(this), next(this), pc(start_pc) { }

  const ScopePtr &Task::begin_scope(Env &env, const ScopePtr &parent) {
    scope = ScopePtr::make(&env.scope_pool, scope, parent);
    return scope;
  }

  void Task::end_scope() {
    auto prev(scope->prev);
    scope->prev = nullptr;
    scope = prev;
  }
}

File: bench/task.py
from bench import bench

def test():
    def task():
        for _ in range(100): yield
        
    t1, t2, t3 = task(), task(), task()
    
    for _ in range(100):
        next(t1)
        next(t2)
        next(t3)

print(int(bench(10000)))
